File: b/lesson7/e/solution.js
const readline = require('readline');

const strToArray = str => str
  .split(' ')
  .filter(Boolean)
  .map(Number);

const rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
});

let n;
const twoPi = 2 * Math.PI;
let r1max = 0;
let r2min = Infinity;
let fs;
let index = 0;

rl.on('line', function (data) {
  const param = data.toString().trim();

  if (!n) {
    n = Number(param);
    fs = Array(2 * n);
    return;
  }

  const [r1, r2, f1, f2] = strToArray(param);
  r1max = Math.max(r1max, r1);
  r2min = Math.min(r2min, r2);

  fs.push({ f: f1, open: true, index });
  fs.push({ f: f2, open: false, index });
  index++;

  if (index >= n) {
    main();
    process.exit(0);
  }
});

const checkResult = () => {
  fs.sort((a, b) => a.f - b.f);

  const opens = new Set();

  let lastOpen;

  fs.forEach(({ open, index, f }) => {
    if (open) {
      opens.add(index);

      if (opens.size === n) {
        lastOpen = f;
      }
    } else {
      opens.delete(index);
    }
  });

  let fres = 0;

  fs.forEach(({ f, open, index }) => {
    if (open) {
      opens.add(index);

      if (opens.size === n) {
        lastOpen = f;
      }
    } else {
      if (opens.size === n) {
        fres += (twoPi + f - lastOpen) % twoPi;
      }
      opens.delete(index);
    }
  });

  let result = fres / 2 * (r2min ** 2 - r1max ** 2);
  console.log(result);
};

const main = () => {
  checkResult();
};

// Алгоритм верный, но на ноде решение не проходит, сдал на пайтоне, 
// причем только на версии 3.9.1

File: b/lesson7/e/solution.py
import math
import pathlib

f = open(pathlib.PurePath(pathlib.Path(__file__).parent.resolve(), 'input.txt'), "r")
n = int(f.readline())

r1max = 0
r2min = 10000000
fs = []

def compare(p):
  return p[0]

for i in range(n):
  r1, r2, f1, f2 = map(float, f.readline().split())
  r1max = max(r1max, r1)
  r2min = min(r2min, r2)

  fs.append([f1, True, i])
  fs.append([f2, False, i])

fs.sort(key=compare)
opens = set()
lastOpen = 0

for i in range(2 * n):
  f, open, index = fs[i]
  if open:
    opens.add(index)

    if len(opens) == n:
      lastOpen = f
  else:
    if index in opens:
      opens.remove(index)

fres = 0

for i in range(2 * n):
  f, open, index = fs[i]
  if open:
    opens.add(index)

    if len(opens) == n:
      lastOpen = f
  else:
    if len(opens) == n:
      fres = fres + (2 * math.pi + f - lastOpen) % (2 * math.pi)

    if index in opens:
      opens.remove(index)

result = fres / 2 * (r2min ** 2 - r1max ** 2)

print(result)
